File: FuturesQuant/src/strategy/cross-correlation-pair.js
/**
 * Cross Correlation Pair indicator implementation
 * Calculates correlation between two assets and generates trading signals based on correlation strength and divergence
 */

class CrossCorrelationPair {
    constructor(options = {}) {
        // Default settings
        this.lookbackPeriod = options.lookbackPeriod || 20;
        this.correlationThreshold = options.correlationThreshold || 0.7;
        this.divergenceThreshold = options.divergenceThreshold || 0.2;
        
        // Initialize storage
        this.data = [];
        this.historicalCorrelation = [];
        this.historicalSpread = [];
    }
    
    update(candleData) {
        // Add candle data to history
        this.data.push(candleData);
        
        // Keep only recent history
        if (this.data.length > this.lookbackPeriod) {
            this.data.shift();
        }
        
        if (this.data.length < this.lookbackPeriod) {
            return {
                correlation: 0,
                spread: 0,
                signal: null,
                strength: 0
            };
        }
        
        // Calculate returns
        const returns1 = this.calculateReturns(this.data.map(d => d.asset1));
        const returns2 = this.calculateReturns(this.data.map(d => d.asset2));
        
        // Calculate correlation
        const correlation = this.calculateCorrelation(returns1, returns2);
        
        // Calculate spread
        const spread = this.data[this.data.length - 1].asset1 - this.data[this.data.length - 1].asset2;
        
        // Store historical values
        this.historicalCorrelation.push(correlation);
        this.historicalSpread.push(spread);
        
        // Keep only recent history
        if (this.historicalCorrelation.length > this.lookbackPeriod) {
            this.historicalCorrelation.shift();
            this.historicalSpread.shift();
        }
        
        // Generate signals based on correlation and divergence
        let signal = null;
        let strength = 0;
        
        if (Math.abs(correlation) >= this.correlationThreshold) {
            // Calculate spread z-score
            const spreadMean = this.calculateMean(this.historicalSpread);
            const spreadStd = this.calculateStandardDeviation(this.historicalSpread);
            const spreadZscore = (spread - spreadMean) / spreadStd;
            
            // Detect divergence
            if (Math.abs(spreadZscore) >= this.divergenceThreshold) {
                if (spreadZscore > 0) {
                    signal = 'SELL';  // Asset1 is overvalued relative to Asset2
                } else {
                    signal = 'BUY';   // Asset1 is undervalued relative to Asset2
                }
                
                // Calculate signal strength based on correlation and divergence
                strength = Math.min(Math.abs(correlation), Math.abs(spreadZscore));
            }
        }
        
        return {
            correlation,
            spread,
            signal,
            strength
        };
    }
    
    calculateReturns(prices) {
        const returns = [];
        for (let i = 1; i < prices.length; i++) {
            returns.push((prices[i] - prices[i-1]) / prices[i-1]);
        }
        return returns;
    }
    
    calculateCorrelation(returns1, returns2) {
        const mean1 = this.calculateMean(returns1);
        const mean2 = this.calculateMean(returns2);
        
        let numerator = 0;
        let denominator1 = 0;
        let denominator2 = 0;
        
        for (let i = 0; i < returns1.length; i++) {
            const diff1 = returns1[i] - mean1;
            const diff2 = returns2[i] - mean2;
            
            numerator += diff1 * diff2;
            denominator1 += diff1 * diff1;
            denominator2 += diff2 * diff2;
        }
        
        return numerator / Math.sqrt(denominator1 * denominator2);
    }
    
    calculateMean(values) {
        return values.reduce((sum, val) => sum + val, 0) / values.length;
    }
    
    calculateStandardDeviation(values) {
        const mean = this.calculateMean(values);
        const squaredDiffs = values.map(val => Math.pow(val - mean, 2));
        return Math.sqrt(this.calculateMean(squaredDiffs));
    }
    
    getHistoricalValues() {
        return {
            correlation: this.historicalCorrelation,
            spread: this.historicalSpread
        };
    }
}

module.exports = CrossCorrelationPair; 

File: FuturesQuant/core/indicator/cross_correlation_pair.py
"""
Cross Correlation Pair indicator implementation.
Calculates correlation between two assets and generates trading signals based on correlation strength and divergence.
"""

import numpy as np
import pandas as pd
from typing import Dict, Any, Tuple
from .base import Indicator

class CrossCorrelationPair(Indicator):
    def __init__(self, 
                 lookback_period: int = 20,
                 correlation_threshold: float = 0.7,
                 divergence_threshold: float = 0.2):
        """
        Initialize the Cross Correlation Pair indicator.
        
        Args:
            lookback_period (int): Period for correlation calculation
            correlation_threshold (float): Minimum correlation strength to consider
            divergence_threshold (float): Threshold for divergence detection
        """
        super().__init__("CrossCorrelationPair", {
            "lookback_period": lookback_period,
            "correlation_threshold": correlation_threshold,
            "divergence_threshold": divergence_threshold
        })
        
        self.lookback_period = lookback_period
        self.correlation_threshold = correlation_threshold
        self.divergence_threshold = divergence_threshold
        
        # Store historical values
        self.historical_correlation = []
        self.historical_spread = []
        
    def calculate(self, data: pd.DataFrame) -> Dict[str, Any]:
        """
        Calculate the cross correlation between two assets and generate signals.
        
        Args:
            data (pd.DataFrame): DataFrame containing price data for both assets
            
        Returns:
            Dict containing:
            - correlation: Current correlation value
            - spread: Current spread between assets
            - signal: Trading signal ('BUY', 'SELL', or None)
            - strength: Signal strength (0-1)
        """
        if len(data) < self.lookback_period:
            return {
                "correlation": 0,
                "spread": 0,
                "signal": None,
                "strength": 0
            }
            
        # Calculate returns
        returns1 = data['asset1'].pct_change()
        returns2 = data['asset2'].pct_change()
        
        # Calculate correlation
        correlation = returns1.rolling(window=self.lookback_period).corr(returns2).iloc[-1]
        
        # Calculate spread
        spread = (data['asset1'] - data['asset2']).iloc[-1]
        
        # Store historical values
        self.historical_correlation.append(correlation)
        self.historical_spread.append(spread)
        
        # Keep only recent history
        if len(self.historical_correlation) > self.lookback_period:
            self.historical_correlation.pop(0)
            self.historical_spread.pop(0)
            
        # Generate signals based on correlation and divergence
        signal = None
        strength = 0
        
        if abs(correlation) >= self.correlation_threshold:
            # Calculate spread z-score
            spread_mean = np.mean(self.historical_spread)
            spread_std = np.std(self.historical_spread)
            spread_zscore = (spread - spread_mean) / spread_std
            
            # Detect divergence
            if abs(spread_zscore) >= self.divergence_threshold:
                if spread_zscore > 0:
                    signal = 'SELL'  # Asset1 is overvalued relative to Asset2
                else:
                    signal = 'BUY'   # Asset1 is undervalued relative to Asset2
                    
                # Calculate signal strength based on correlation and divergence
                strength = min(abs(correlation), abs(spread_zscore))
                
        return {
            "correlation": correlation,
            "spread": spread,
            "signal": signal,
            "strength": strength
        }
        
    def get_historical_values(self) -> Tuple[list, list]:
        """
        Get historical correlation and spread values.
        
        Returns:
            Tuple of (correlation_history, spread_history)
        """
        return self.historical_correlation, self.historical_spread 